Zend/zend_execute.c

```diff
@@ -770,6 +799,97 @@ static int zend_verify_internal_arg_type(zend_function *zf, uint32_t arg_num, zv
 	return 1;
 }
 
+static inline zend_bool zend_verify_scalar_property_type(zend_uchar type, zval *property, zend_bool strict) {
+	if (EXPECTED(!strict)) {
+		if (UNEXPECTED(!ZEND_SAME_FAKE_TYPE(type, Z_TYPE_P(property)))) {
+			return zend_verify_weak_scalar_type_hint(type, property);
+		}
+		return 1;
+	}
+
+	if (UNEXPECTED(type == IS_DOUBLE && Z_TYPE_P(property) == IS_LONG)) {
+		double dest = (double) Z_LVAL_P(property);
+		zval_ptr_dtor(property);
+		ZVAL_DOUBLE(property, dest);
```

Comment:
Updating source in-place is not allowed. Source may be kept in read-only memory.